### PR TITLE
Tree accessors and random generator

### PR DESCRIPTION
**Pull request by imane-aanna**
This branch aims to add accessors to the tree, and create a random generator of the tree.

### CODE REVIEW

File: code/Tree.cs [hunk 1]
```diff
@@ -37,15 +37,90 @@ public class Tree
...
+        /// <param name="verticeId"> The id of the vertice parent. </param>
+        /// <returns> Returns the number of children vertices (0 if there are none).
+        /// If the vertice parent does not exist, it returns -1 .</returns>
+        public int NbChildren(int verticeId)
+        {
+            int nbChildren = 0;
+            Dictionary<int, List<int>> temporary = children;
+
+            if (!temporary.ContainsKey(verticeId))
+                nbChildren = -1;
```
Comment (pradal) [inline code/Tree.cs]: NbChildren has to be >= 0
In the case of the vertex do not belong to the Tree, return 0 or an error.
I would prefer 0.

**Comment by imane-aanna** [comment]
@pradal : 
New changes have been added to this pull request:
- The returned values in case of errors are now changed.
- Unit tests have been created to validate the random tree generator.
- The function that enables to remove a vertex has been added.

File: code/Tree.cs [hunk 2]
```diff
@@ -37,15 +38,90 @@ public class Tree
...
             return count;
         }
 
         /// <summary>
+        /// Counts the number of vertices in the tree.
+        /// </summary>
+        /// <returns> Returns the total number of vertices. </returns>
+        public int NbVertices()
+        {
+            int nbVertices = parent.Count();
```
Comment (pradal) [inline code/Tree.cs]: @imane-aanna : Use the Count method instead

File: code/Tree.cs [hunk 3]
```diff
@@ -54,7 +130,7 @@ private int NewId()
             {
                 id++;
             }
-            while (children.ContainsKey(id));
+            while (parent.ContainsKey(id));
```
Comment (pradal) [inline code/Tree.cs]: @imane-aanna 
The complexity is O(n).
You can also use something like:
 max(parent)+1

File: code/Tree.cs [hunk 4]
```diff
@@ -125,10 +202,101 @@ void ReplaceParent(int parentId, int childId)
...
+                }
+            }
+        }
+
+        /// <summary>
+        ///  Generates a random tree with a specified number of vertices and children per vertex
+        /// </summary>
+        /// <param name="nbVertices"> The number of tree's vertices </param>
+        /// <param name="nbChildren"> The maximum number for a vertex </param>
+        /// <returns> Returns the generated tree </returns>
```
Comment (pradal) [inline code/Tree.cs]: @imane-aanna : I will add this algorithm outside this class in another module
like algorithm.cs

File: code/Tree.cs (hunk 2 above)
Comment (imane-aanna) [inline code/Tree.cs]: I wrote in my side notes that it was a duplicate of the function Count() and I wanted to ask you about it indeed.

Comment (imane-aanna) [inline code/Tree.cs]: @pradal : I wrote in my side notes that it was a duplicate of the function Count() and I wanted to ask you about it indeed.

**Comment by imane-aanna** [comment]
@pradal : Your remarks have been taken into account. In the last commit, I : 
- Deleted NbVertices() and kept Count().
- Changed the way a new id is attributed.
- Placed the random tree generator in a new Algorithm's class.

**Comment by pradal** [comment]
Do not remove NbVertices function.
Just add
int NbVertices()
{ return Count(); }

**Comment by imane-aanna** [comment]
@pradal  : Could you please check if everything is fine so that the branch is merged with the master and a new one is created for the property tree ?

**Comment by pradal** [comment]
Good. I merge it

**Pull request merged**